fork-management-for-infra-providers/README.md

```diff
@@ -0,0 +1,148 @@
+  - https://github.com/giantswarm/cluster-api-provider-gcp
+
+1. A repository for bundling the operator as an app 
+- Naming: https://github.com/giantswarm/\<fork-name\>-app
+- Examples
+  - https://github.com/giantswarm/cluster-api-app
+  - https://github.com/giantswarm/cluster-api-provider-aws-app
+  - https://github.com/giantswarm/cluster-api-provider-cloud-director-app
+  - https://github.com/giantswarm/cluster-api-provider-gcp-app
+
+### Syncing fork repositories
+
+There is no common way.
+
+In CAPVCD, since we have some diffs between our fork and upstream repo, we have a mechanism as follows:
+- `main` branch of the repo is the same as the upstream one. We sync this branch by clicking `Sync fork` button in the UI.
+- `giantswarm-main` branch of the repo contains our changes. We manually rebase it on top of `main` branch when necessary and open a new PR each time. The developer is supposed to resolve conflicts during rebasing.
+ 
+### Bundling + Customization
+
+There are some differences between repositories but the general structure in app repositories is as follows:
+
+- We have Makefile + Bash scripts to
+  - Check out a specific commit/branch of fork/upstream repo. [Example](https://github.com/giantswarm/cluster-api-provider-cloud-director-app/blob/adfc74115d9b7aeda60d885500122be3fb434d2e/Makefile.custom.mk#L14)
+  - Generate base manifests by running scripts of the core repo
+  - Customize base manifests with kustomization. [Example](https://github.com/giantswarm/cluster-api-provider-cloud-director-app/blob/adfc74115d9b7aeda60d885500122be3fb434d2e/Makefile.custom.mk#L19)
+  - Update manifests under `helm` folder for the app. [Example](https://github.com/giantswarm/cluster-api-provider-cloud-director-app/blob/adfc74115d9b7aeda60d885500122be3fb434d2e/Makefile.custom.mk#L28)
+
+When it is necessary to update the app content, developers are supposed to
+- run make targets manually to update manifests and values in the app repo
+- open a PR in the app repo
+
+### Image Management
+
+For some providers, we retag upstream images with retagger.
+For some providers like CAPVCD, we build/push images manually.
+
+## Design proposals
+
+### 1. Repo management
+
+#### 1.a Separate repositories for fork and app
+
+We can have two repositories per operator. One is only fork of upstream and the other one is for only bundling it as an app.
+
+#### 1.b One repository for fork and app
+
+We can maintain bundling and customization logic in the same repo.
+
+#### Decision
+
+- `1.a` is the current situation.
+- `1.b` will lead to many conflicts with the upstream repository and it will be hard to automate conflict resolution.
+
+`1.a` is decided.
+
+---
+
+### 2. Branching strategy in fork repo
+
+#### 2.a main vs giantswarm-main
+
+We can maintain two branches in the fork repo. 
+- `main` will be just the same as `main` of upstream repository. This branch can be synced automatically.
+- `giantswarm-main` will contain our changes if necessary. 
```

Comment:
What is the value of having a branch which is upstream `main` with our changes, what would we use it for?

Reply:
From upstream docs https://github.com/kubernetes-sigs/cluster-api/blob/main/CONTRIBUTING.md#branches

> Cluster API has two types of branches: the main branch and release-X branches.
>
> The main branch is where development happens. All the latest and greatest code, including breaking changes, happens on main.
>
> **The release-X branches contain stable, backwards compatible code.** On every major or minor release, a new branch is created. It is from these branches that minor and patch releases are tagged. In some cases, it may be necessary to open PRs for bugfixes directly against stable branches, but this should generally not be the case.

We should never deploy a version of controller that is coming from upstream main branch, because it is not considered stable, it is latest, greatest and often breaking code. Therefore we should really not care about maintaining our changes in some main branch in our fork (named `giantswarm-main` here), as it would be additional maintenance work.

The changes in our fork, when we really must implement them, should always come on top of stable upstream code, which means that we would do PRs to release branches directly.

If some custom change in our fork gets to be long(er) lived (we should not have many of those, but about that in another comment) because it takes more time to get it released in upstream, and there is a new release, meaning there is a new release branch, then those custom changes would be cherry-picked from old release branch to a newer release branch.

Reply:
In theory I would agree that our forks should always come out of an  upstream release branch. However this is against the experience we had with CAPO. Upstream was not releasing for many month. We even had several PRs merged into it, but it was not released for a very long time. So upstream `main` was already `giantswarm-main` in a way. 
I don`t think backporting all this changes to a released version would be worth the effort.

Reply:
I think it should be up to the PEs to decide which branch is the most stable and best base for a fork.

Reply:
But wouldn't those unreleased changed go into a `release-(n+1)` branch?

For example, in CAPA that have a `release-2.0` and a `release-1.5`. The `release-2.0` is what new releases are tagged off of, so `v2.0.2` being the latest from that branch, and the `release-1.5` is for any backported features with `v1.5.2` being the latest release from that branch. Once new breaking changes are planned a `release-2.1` (or whatever) branch will be created first. This is the approach I've seen for at least CAPI & CAPA and I think CAPZ. 

If the upstream processes change that much between providers we then need to call that out in this and plan accordingly.

Reply:
> I think it should be up to the PEs to decide which branch is the most stable and best base for a fork.

This is documented in upstream projects, and unless some other projects are for some reason different, the stable branches used for releases are `release-xxx` branches, and `main` is a development branch.

Reply:
> However this is against the experience we had with CAPO. Upstream was not releasing for many month. We even had several PRs merged into it, but it was not released for a very long time. So upstream main was already giantswarm-main in a way.

This sounds like an exception, as it seems that a large amount of work, features and fixes got accumulated in `main` branch that was not released at all, so backporting that to the older version would not make sense in that case, as new release should be released.

Relatively active upstream CAPI projects do not have this issue, I think this is in exception rather than a rule.

In this case we should push upstream to cut a release, and if the project is inactive/unmaintained that much in general to the point that we are always waiting too much for new minor release (to get those accumulated features released), we should get involved in upstream project more directly and get somebody from Giant Swarm to become reviewer/maintainer and drive the project forward. We already have folks that became reviewers/maintainers in other upstream projects.

Reply:
> This sounds like an exception, as it seems that a large amount of work, features and fixes got accumulated in main branch that was not released at all, so backporting that to the older version would not make sense in that case, as new release should be released.

In this case, where new release has not been cut for that long, the `main` branch state is probably what the latest `release-xxx` branch should be, but the project is inactive/slow, so that work is not done.